src/Http/Http/src/IHttpContextContainer.cs

```diff
@@ -6,6 +6,6 @@ namespace Microsoft.AspNetCore.Http
 {
     public interface IHttpContextContainer
     {
-        HttpContext HttpContext { get; }
+        DefaultHttpContext HttpContext { get; }
```

Comment:
Is this going to be a problem if you seal `DefaultHttpContext`?

Reply:
Ehhhh, maybe? For that 1% that want to make a derived version. This was to avoid the casting in the default HttpContextFactory

Reply:
Ah, so `HttpContext` can still be overriden an flow, it isn't bonded to `DefaultHttpContext`?

(Otherwise just rename `DefaultHttpContext` to `HttpContext`)

Reply:
The scenario here is you want to implement a server with a custom lifetime management of the HttpContext that isn't a DefaultHttpContext. This change ties you to DefaultHttpContext. The alternative would be to cast detect the cast.

The another approach would be to only assign this field once since it can't change per request anyways (that would likely mitigate the cost of the downcast).

Reply:
This is what it would look like.

```C#
private HttpContext CreateHttpContext(IFeatureCollection featureCollection)
{
    if (featureCollection is IHttpContextContainer container)
    {
        var context = container.HttpContext;
        if (context is DefaultHttpContext defaultHttpContext)
        {
            defaultHttpContext.FormOptions = _formOptions;
        }
        else
        {
            var formFeature = new FormFeature(context.Request, _formOptions);
            featureCollection.Set<IFormFeature>(formFeature);
        }
    }

    return new DefaultHttpContext(featureCollection)
    {
        FormOptions = _formOptions
    };
}
```

Reply:
Perfect :)

Reply:
Need to measure the performance here, I think this will regress stuff...

Reply:
As this didn't happen, will it before 3.0 rtm?

If not, `IHttpContextContainer` which currently is part of public api will only allow one flavor of `HttpContext`? 
I wouldn't say you can reduce all reasons for wanting extensibility to just 'different lifetime'. If you want to add functionality that isn't on `DefaultHttpContext` — while keeping lifetime management as is — we're toast coming 3.0? 

With this you'd be in for some nasty surprises trying to thread your own `HttpContext` through. 
When important apis get very concrete shapes you might as well just stop allowing any extensibility of `HttpContext` ¯\\_(ツ)_/¯.

Either way, extensible or not, this should be deliberate and well documented decision for 3.0.

Reply:
It's unlikely to happen, though I do want to hide the interface a little and maybe rename it to IDefaultHttpContextContainer or something similar.

> I wouldn't say you can reduce all reasons for wanting extensibility to just 'different lifetime'. If you want to add functionality that isn't on DefaultHttpContext — while keeping lifetime management as is — we're toast coming 3.0?

You can *always* write your own factory, and can count the number of servers that would use this feature on one hand so it's not a big deal in reality.

> With this you'd be in for some nasty surprises trying to thread your own HttpContext through.
When important apis get very concrete shapes you might as well just stop allowing any extensibility of HttpContext ¯\(ツ)/¯.

Can you elaborate?

> Either way, extensible or not, this should be deliberate and well documented decision for 3.0.

 What exactly?

Reply:
> With this you'd be in for some nasty surprises trying to thread your own `HttpContext` through.

You'd have to create your own `HttpContextFactory` for that, which would be creating its own `HttpContext`s so would be unaffected?

Reply:
That's right. If you want to make a different type of HttpContext then you need your own factory and now you control the lifetime.

Reply:
It won't be strictly equivalent though as `HttpProtocol` implements that interface `IDefaultHttpContextContainer` alone not for my HttpContext. Not sure what it's being used for or will be used for but just creating your own factory is already not enough to get full equivalence.

Reply:
I'm not sure what you mean, or how you plan to use this API. Did you write a custom server? Or is this strictly hypothetical?

Reply:
I'm asking whether creating a factory is enough to get full equivalence, why is HttpProtocol specifically implementing IDefaultHttpContextContainer?

Reply:
> I'm asking whether creating a factory is enough to get full equivalence

I don't even know what you're asking for? equivalence of what? 

> why is HttpProtocol specifically implementing IDefaultHttpContextContainer

- https://github.com/aspnet/AspNetCore/commit/688ad19170df1b41ca6539805fd9348bf9eb5dac
- https://github.com/aspnet/AspNetCore/commit/a11c68ae35547d2db9b4fbb6c76d695758301b2c
- https://github.com/aspnet/AspNetCore/commit/ea344bf726f73d9928c03b613386ee78cfe40650

Reply:
>I don't even know what you're asking for? equivalence of what?

My ask is simple: What if I want equivalent reusability semantics to DefaultHttpContext, but have reasons to create my own HttpContext fields.

Now DefaultHttpContext is sealed, what to do? 

If it's possible — without having to create my own Kestrel protocol too — then there's full equivalence between DefaultHttpContext and my own HttpContext. 
If it's not possible then there's only equivalence that comes at a performance penalty.

I think that's an important aspect to highlight with respect to these changes.

Reply:
> My ask is simple: What if I want equivalent reusability semantics to DefaultHttpContext, but have reasons to create my own HttpContext fields.
> Now DefaultHttpContext is sealed, what to do?

I think it's fine to then file a bug and request it be unsealed. Until then it's sealed because nobody has done that (or not enough that I think it matters). When the masses complain about this change then we'll have some data we can use to unseal the type. Thus far my theory is that 0.01% of people do this so it won't affect anymore in real life.

> I think that's an important aspect to highlight with respect to these changes.

 The change was marked as breaking and the relevant parties know and hopefully, people try out the previews and give feedback with concrete scenarios, not hypothetical ones.

Reply:
But wasn't there a straightforward fix for HttpContextContainer to support HttpContext instead? 

All I'm saying, if it's low effort/trade-off to keep the full equivalence then that would be my preferred option.

Reply:
> But wasn't there a straightforward fix for HttpContextContainer to support HttpContext instead?
> All I'm saying, if it's low effort/trade-off to keep the full equivalence then that would be my preferred option.

More complicated code that i'm not willing to write because of the narrow-ness of the scenario. This is a new extensibility point introduced in 3.0 that only one of our servers supports. If you have a concrete scenaro, file an issue with a concrete use case and we'll evaluate it.